### PR TITLE
Add filter text box to new ride selection window

### PR DESCRIPTION
**Pull request by trzejos**
Adds a text box and clear button to the new ride window allowing users to filter based on the following:
- Ride Type
- Ride/Vehicle Name
- Author Name
- Identifier Name (rct_object_entry Name field)
- Object Repository Item Filename

Also fixes a bug in the Object Selection window where backspacing over your filter string will not update the search string if you remove all characters.

Examples:
image
image
image
image

### CODE REVIEW

**Comment by trzejos** [comment]
@Gymnasiast if this last build completes successfully, I don't think I'm making any more changes without feedback and this can be taken out of draft state

File: src/openrct2-ui/windows/EditorObjectSelection.cpp [hunk 1]
```diff
@@ -793,7 +793,7 @@ class EditorObjectSelectionWindow final : public Window
 
     void OnTextInput(WidgetIndex widgetIndex, std::string_view text) override
     {
-        if (widgetIndex != WIDX_FILTER_TEXT_BOX || text.empty())
+        if (widgetIndex != WIDX_FILTER_TEXT_BOX)
```
Comment (Broxzier) [inline src/openrct2-ui/windows/EditorObjectSelection.cpp:796]: What's the reason for this change?

**Comment by karst** [comment]
Don't forget about a changelog entry. I think this probably would fit best under "Improved".

File: src/openrct2-ui/windows/EditorObjectSelection.cpp (hunk 1 above)
Comment (trzejos) [reply src/openrct2-ui/windows/EditorObjectSelection.cpp:796]: You can test this yourself in an official build. In the object selection window, just enter a single character into the filter text box. You will notice the handle on the scroll bar expands, indicating that items have been filtered from the list. If you then hit backspace, the text box will now be empty, but the scroll bar handle will not shrink back to it's original size. The old filter string is still being used and will continue to be used until either a different string is entered or the clear button is pressed.

Comment (trzejos) [reply src/openrct2-ui/windows/EditorObjectSelection.cpp:796]: Same bug in `TrackList.cpp`

Comment (Gymnasiast) [reply src/openrct2-ui/windows/EditorObjectSelection.cpp:796]: This belongs in a PR of its own.

File: distribution/changelog.txt [hunk 2]
```diff
@@ -10,12 +10,14 @@
...
 - Improved: [#18970] Trying to load a non-park save will now display a context error.
 - Improved: [#19044] Added special thanks to RMC and Wiegand to the About page.
+- Improved: [#19067] New Ride window now allows filtering similarly to Object Selection.
 - Change: [#19018] Renamed actions to fit the naming scheme.
 - Fix: [#18467] “Selected only” Object Selection filter is active in Track Designs Manager, and cannot be toggled.
 - Fix: [#18905] Ride Construction window theme is not applied correctly.
 - Fix: [#18911] Mini Golf station does not draw correctly from all angles.
 - Fix: [#18971] New Game does not prompt for save before quitting.
 - Fix: [#19026] Park loan is clamped to a 32-bit integer.
+- Fix: [#19067] Object selection filter is updated correctly when set to empty string.
```
Comment (Gymnasiast) [inline distribution/changelog.txt:20]: When you move this fix to its own PR (see the other line note), please reword. The changelog and commit message should list the problem that was solved, rather than the solution.

File: src/openrct2-ui/windows/EditorObjectSelection.cpp (hunk 1 above)
Comment (trzejos) [reply src/openrct2-ui/windows/EditorObjectSelection.cpp]: I've removed this change and opened #19112 to track this bug

File: distribution/changelog.txt (hunk 2 above)
Comment (trzejos) [reply distribution/changelog.txt]: This has been removed along with the corresponding code change

File: src/openrct2-ui/windows/NewRide.cpp [hunk 3]
```diff
@@ -698,6 +741,55 @@ class NewRideWindow final : public Window
...
+    bool IsFiltered(rct_ride_entry* rideEntry)
+    {
+        if (Filter.empty())
+            return true;
+
+        return IsFilterInRideType(rideEntry) || IsFilterInRideName(rideEntry) || IsFilterInIdentifier(rideEntry)
+            || IsFilterInAuthors(rideEntry) || IsFilterInFilename(rideEntry);
+    }
+
+    bool IsFilterInRideType(rct_ride_entry* rideEntry)
```
Comment (Gymnasiast) [inline src/openrct2-ui/windows/NewRide.cpp:753]: Please use `const rct_ride_entry&` here and in the other functions you add. We’re trying to move away from pointers in function signatures.

File: src/openrct2-ui/windows/NewRide.cpp [hunk 4]
```diff
@@ -45,6 +46,7 @@ static constexpr const int32_t WindowWidth = 601;
 static constexpr const int32_t RideListItemsMax = 384;
 static constexpr const int32_t RideTabCount = 6;
 static constexpr const int32_t GroupByTrackTypeWidth = 172;
+static std::string Filter;
```
Comment (Gymnasiast) [inline src/openrct2-ui/windows/NewRide.cpp:49]: For static variables like this one, we use the notation `_filter`.

**Comment by Gymnasiast** [comment]
Hey, sorry for the feedback coming in very small chunks. I now left all my review bits, but since my knowledge of string handling is somewhat limited, I’m trying to see if I can get another team member to take a look at that.

File: src/openrct2-ui/windows/NewRide.cpp [hunk 5]
```diff
@@ -286,6 +292,10 @@ class NewRideWindow final : public Window
         PopulateRideList();
         InitScrollWidgets();
 
+        Filter.resize(MAX_PATH);
+        widgets[WIDX_FILTER_TEXT_BOX].string = Filter.data();
+        Filter.clear();
```
[the remark this answers is not in the document]
Comment (trzejos) [reply src/openrct2-ui/windows/NewRide.cpp:297]: I wanted to avoid the string getting resized after giving the textbox widget a pointer to its data.

File: src/openrct2-ui/windows/NewRide.cpp (hunk 3 above)
Comment (trzejos) [reply src/openrct2-ui/windows/NewRide.cpp:753]: Using `const rct_ride_entry&` is a little troublesome because the call to `get_ride_naming` in `ISFilterInRideType` takes a `rct_ride_entry*`. I could make a copy and pass that instead or I could just remove the `const` and use `rct_ride_entry&` instead. I'm not sure what your preference is though.

File: src/openrct2-ui/windows/NewRide.cpp (hunk 4 above)
Comment (trzejos) [reply src/openrct2-ui/windows/NewRide.cpp]: This has been changed

File: src/openrct2-ui/windows/NewRide.cpp (hunk 3 above)
Comment (Gymnasiast) [reply src/openrct2-ui/windows/NewRide.cpp]: It’s fine, it’s not that intrusive.

**Comment by Gymnasiast** [comment]
The PR seems to work, though I’m not 100% sure if should use the same criteria as the object selection. For example, authors and identifiers are not terribly relevant or even visible in the New Ride window. Conversely, it doesn’t search in the description, which is something I would expect in the New Ride window.

Before you change anything, let’s first hear what other people think.

**Comment by trzejos** [comment]
I think @spacek531 was asking for identifier searching in Discord, but that may have been in reference to scenery rather than rides. I also considered searching the description, but I don't really have a strong feeling on it one way or another. It wouldn't be difficult to add at this point, so I agree we can wait and see if others have any input.

File: src/openrct2-ui/windows/NewRide.cpp [hunk 6]
```diff
@@ -286,6 +292,10 @@ class NewRideWindow final : public Window
         PopulateRideList();
         InitScrollWidgets();
 
+        _filter.resize(MAX_PATH);
+        widgets[WIDX_FILTER_TEXT_BOX].string = _filter.data();
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/NewRide.cpp:296]: Instead of doing this set the  `widgets[WIDX_FILTER_TEXT_BOX].string` in the `OnPrepareDraw` function. That is guaranteed to be called prior to drawing so the pointer will be valid at time of drawing. Previously we used to do dynamic strings via a similar method with the common format args but its also pretty messy.

File: src/openrct2-ui/windows/NewRide.cpp [hunk 7]
```diff
@@ -45,6 +46,7 @@ static constexpr const int32_t WindowWidth = 601;
 static constexpr const int32_t RideListItemsMax = 384;
 static constexpr const int32_t RideTabCount = 6;
 static constexpr const int32_t GroupByTrackTypeWidth = 172;
+static std::string _filter;
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/NewRide.cpp:49]: This should become a member variable of the window. You are clearing it in `OnOpen` so its not like it needs to be a local global variable (unless I'm not understanding somthing)

**Review by duncanspumpkin — CHANGES_REQUESTED** [review]
Good work. A few little tweaks still needed to make this just right.

File: src/openrct2-ui/windows/NewRide.cpp (hunk 7 above)
Comment (trzejos) [reply src/openrct2-ui/windows/NewRide.cpp:49]: I think you're right. I first attempted this when I had no knowledge about openrct2 code and was primarily mirroring the ObjectSelection window. I can also switch to u8string as well.

**Comment by trzejos** [comment]
I think this has all the tweaks you requested now @duncanspumpkin

**Review by duncanspumpkin — APPROVED** [review]
All works quite well.

**Comment by duncanspumpkin** [comment]
Looks like you'll need to rebase to fix the chanelog conflict.
